### ISSUE TITLE
Font banks! And/or sensible RUN-STOP+RESTORE reset behavior for FONT/CHARSET

### ISSUE THREAD

**Issue by dansanderson** [MEGA65/mega65-rom-public#16]
**Problem**
On MEGA65, RUN-STOP+RESTORE does not reset anything related to the character glyphs set by `CHARDEF`. If a BASIC program replaces character graphics in a way that makes immediate mode commands difficult to enter then exits abnormally, the user may expect RUN-STOP+RESTORE to return the editor to a usable state (colors, clear screen, modes). With replaced glyphs, the user would have to know to type `FONT C`—possibly blindly—to fully restore the screen.

A user might especially expect the font to reset if they're used to a C64, which resets custom fonts by resetting the ROM bank switching used for custom characters. Notably, the C64 retains the custom font in RAM in this case.

**Modest solution: Automatic FONT C**
One solution is for RUN-STOP+RESTORE to perform the equivalent of `FONT C` as one of its reset behaviors.

It is possible that a user might be using `FONT A` or `FONT B` as a way to set the font of their editor experience, and not just within a program. These users would have to re-type `FONT A` or `FONT B` after RUN-STOP+RESTORE. A possible improvement: remember the most recent `FONT` command, and use that font as the reset font for RUN-STOP+RESTORE.

It is possible that a user might be using `CHARSET` to define a custom font for their editor experience, e.g. from an AUTOBOOT.C65 file. We may be able to leave this use case aside. Or we can consider a more ambitious solution.

**Ambitious solution: font banks!**
A font bank is a character set in memory assigned to a label. The current font bank is the one actively used to render characters. `FONT <b>` switches the current font bank to the font bank with that label.

A font bank with a single raw letter name is a read-only system font, such as A, B, or C. D-Z are reserved for future system expansion. A font bank with a numeric label is an allocated writeable font. All numeric labels begin uninitialized, with no memory allocated. When the current font switches to a writeable font bank that has not been initialized, it is allocated and initialized with the contents of the previously selected font bank. An allocated writeable bank remains in memory when the current font is changed.

`CHARSET` writes to the currently selected font bank. If the current font bank is a read-only bank, it copies the current bank to bank 0, switches to bank 0, then makes its update. (This is for backwards compatibility. Alternatively, it could just throw an error if the current bank is not writeable.)

`FONT CLR` deallocates the current font bank (if writeable) then switches to bank C. `FONT LOAD <filename> [,D<drive>] [,U<unit>]` loads font data from a file into the current bank. `FONT SAVE <filename>` saves font data from the current bank to a file. (Maybe `FONT LOAD` and `FONT SAVE` accept a `[,B<bank>]` parameter. Maybe `LOAD` when the current bank is read-only switches to bank 0 first.)

`FONT COPY <b1> TO <b2> [,<n>...]` copies the contents of one font bank to another; `b2` must be writeable. If one or more screen codes `n` are provided, only those screen codes are copied.

All commands that accept a bank parameter also accept a parenthesized expression that evaluates to the bank label. (Only for numbers? Numbers and strings?)

In this solution, RUN-STOP+RESTORE switches to font bank C without loss of data. The user can restore any active font bank with the `FONT` command.

**Comment by Edilbert** [comment]
Resetting the font to default after STOP RESTORE is probably a useful feature and is easy to implement.
Also a command FONT LOAD fits well, which loads a font from file and uses it immediately.
The other suggestions however would need a lot of ROM code and RAM reservation for font management.
I propose, that I put STOP RESTORE sets default font and FONT LOAD on my TODO list.
Programs, that use more than one font could store them in attic RAM and transfer them to the VIC font address via EDMA command.

**Comment by dansanderson** [comment]
Sounds good!

I imagine a BSAVE with the right arguments would be the equivalent of FONT SAVE, the reverse of a FONT LOAD? Would BLOAD with those same arguments be the equivalent of FONT LOAD, or does FONT LOAD have more to do?

**Comment by dansanderson** [comment]
I wrote this suggestion before I understood that CHARDEF/FONT are using the writability of the VIC-IV character generator addresses to update glyphs. As of now, the C65 charset ROM data is the only source of truth for this data after the VIC's internal character generator memory has been modified. For a different purpose, I proposed that the VIC expose its own copy of "font C" in a new read-only address region, so it is accessible even if C65 ROM has been overwritten by a program (so Freezer et al can disable and restore a custom charset). https://github.com/MEGA65/mega65-core/issues/667

Without a Hypervisor/core change, the Run/Stop+Restore proposal would be dependent on "font C" C64 charset ROM remaining in memory (as the `FONT C` command necessarily does). This might be fine: perhaps Run/Stop+Restore is best considered a feature of the kernel, and anyone overwriting ROM regions should not expect those features to work.

(It's worth noting that Paul is inclined to discourage writing to the VIC character generator memory in our documented API surfaces. CHARSET/FONT do this today. To not do this, we'd have to adopt something more like the "font bank" feature I describe above, where BASIC reserves VIC-accessible memory for custom character sets. I don't know if it's possible to crawl back the writable charset addresses; I suspect we already have major titles using it.)
